File: src/main/java/io/thegamingmahi/luacord/environment/wrappers/storage/JsonStorage.java
package io.thegamingmahi.luacord.environment.wrappers.storage;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.stream.JsonReader;
import io.thegamingmahi.luacord.Utilities;
import io.thegamingmahi.luacord.environment.exception.StorageObjectException;
import io.thegamingmahi.luacord.environment.plugin.LukkitPlugin;
import org.luaj.vm2.LuaBoolean;
import org.luaj.vm2.LuaString;
import org.luaj.vm2.LuaValue;
import org.luaj.vm2.lib.jse.CoerceJavaToLua;

import java.io.*;

/**
 * @author jammehcow
 */

public class JsonStorage extends StorageObject {
    private Gson gson = new GsonBuilder().create();
    private JsonObject object;

    public JsonStorage(LukkitPlugin plugin, String path) {
        super(plugin, path, Storage.JSON);

        try {
            this.object = new JsonParser().parse(new JsonReader(new FileReader(this.getStorageFile()))).getAsJsonObject();
        } catch (FileNotFoundException | IllegalStateException e) {
            this.object = new JsonObject();
        }
    }

    @Override
    protected LuaBoolean exists(String path) {
        return object.has(path) ? LuaValue.TRUE : LuaValue.FALSE;
    }

    @Override
    public LuaBoolean setDefaultValue(LuaString path, LuaValue value) throws StorageObjectException {
        if (object.has(path.checkjstring())) {
            setValue(path, value);
            return LuaValue.TRUE;
        }
        return LuaValue.FALSE;
    }

    @Override
    public void setValue(LuaString path, LuaValue value) throws StorageObjectException {
        object.add(path.checkjstring(), gson.toJsonTree(Utilities.getObjectFromLuavalue(value)));
    }

    @Override
    protected LuaBoolean clearVaule(LuaString path) throws StorageObjectException {
        if (object.has(path.checkjstring())) {
            object.remove(path.checkjstring());
            return LuaValue.TRUE;
        }
        return LuaValue.FALSE;
    }

    @Override
    public LuaValue getValue(LuaString path) throws StorageObjectException {
        return CoerceJavaToLua.coerce(object.get(path.checkjstring()));
    }

    @Override
    public void save() {
        this.preSaveCheck();
        try {
            Writer writer = new FileWriter(this.getStorageFile());
            gson.toJson(this.object, writer);
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/io.thegamingmahi.luacord/environment/wrappers/storage/JsonStorage.kt
package io.thegamingmahi.luacord.environment.wrappers.storage

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.stream.JsonReader
import io.thegamingmahi.luacord.Utilities
import io.thegamingmahi.luacord.environment.exception.StorageObjectException
import io.thegamingmahi.luacord.environment.plugin.LukkitPlugin
import org.luaj.vm2.LuaBoolean
import org.luaj.vm2.LuaString
import org.luaj.vm2.LuaValue
import org.luaj.vm2.lib.jse.CoerceJavaToLua
import java.io.*

/**
 * JSON storage implementation
 *
 * @author jammehcow
 * @author TheGamingMahi (Kotlin rewrite)
 */
class JsonStorage(plugin: LukkitPlugin, path: String) : StorageObject(plugin, path, Storage.JSON) {

    private val gson: Gson = GsonBuilder().create()
    private var jsonObject: JsonObject

    init {
        jsonObject = try {
            JsonParser().parse(JsonReader(FileReader(getStorageFile()))).asJsonObject
        } catch (e: FileNotFoundException) {
            JsonObject()
        } catch (e: IllegalStateException) {
            JsonObject()
        }
    }

    override fun exists(path: String): LuaBoolean {
        return if (jsonObject.has(path)) LuaValue.TRUE else LuaValue.FALSE
    }

    @Throws(StorageObjectException::class)
    override fun setDefaultValue(path: LuaString, value: LuaValue): LuaBoolean {
        return if (jsonObject.has(path.checkjstring())) {
            setValue(path, value)
            LuaValue.TRUE
        } else {
            LuaValue.FALSE
        }
    }

    @Throws(StorageObjectException::class)
    override fun setValue(path: LuaString, value: LuaValue) {
        jsonObject.add(path.checkjstring(), gson.toJsonTree(Utilities.getObjectFromLuavalue(value)))
    }

    @Throws(StorageObjectException::class)
    override fun clearValue(path: LuaString): LuaBoolean {
        return if (jsonObject.has(path.checkjstring())) {
            jsonObject.remove(path.checkjstring())
            LuaValue.TRUE
        } else {
            LuaValue.FALSE
        }
    }

    @Throws(StorageObjectException::class)
    override fun getValue(path: LuaString): LuaValue {
        return CoerceJavaToLua.coerce(jsonObject.get(path.checkjstring()))
    }

    override fun save() {
        preSaveCheck()
        try {
            FileWriter(getStorageFile()).use { writer ->
                gson.toJson(jsonObject, writer)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}